src/wallet/wallet_rpc_server.cpp

```diff
@@ -2431,6 +2431,97 @@ bool wallet_rpc_server::on_create_wallet(const wallet_rpc::COMMAND_RPC_CREATE_WA
 	return true;
 }
 //------------------------------------------------------------------------------------------------------------------------------
+bool wallet_rpc_server::on_restore_wallet(const wallet_rpc::COMMAND_RPC_RESTORE_WALLET::request &req, wallet_rpc::COMMAND_RPC_RESTORE_WALLET::response &res, epee::json_rpc::error &er)
+{
+	if(m_wallet_dir.empty())
+	{
+		er.code = WALLET_RPC_ERROR_CODE_NO_WALLET_DIR;
+		er.message = "No wallet dir configured";
+		return false;
+	}
+
+	std::string wallet_file = m_wallet_dir + "/" + req.filename;
+
+	std::string seed = req.seed;
+
+	std::vector<std::string> wseed;
+	boost::algorithm::trim(seed);
+	boost::split(wseed, req.seed, boost::is_any_of(" "), boost::token_compress_on);
+
+	std::string language;
+	crypto::secret_key_16 seed_14;
+	uint8_t seed_extra;
+	crypto::secret_key seed_25;
+	bool decode_14 = false, decode_25 = false;
+	if(wseed.size() == 12 || wseed.size() == 14)
+	{
+		decode_14 = crypto::Electrum14Words::words_to_bytes(req.seed, seed_14, seed_extra, language);
+	}
+	else if(wseed.size() >= 24 && wseed.size() <= 26)
+	{
+		decode_25 = crypto::Electrum25Words::words_to_bytes(req.seed, seed_25, language);
+	}
+	else
+	{
+		er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+		er.message = "Unkown seed size " + std::to_string(wseed.size()) + " please enter 12, 14, 24, 25 or 26 words";
+		return false;
+	}
+
+	if(!decode_14 && !decode_25)
+	{
+		er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+		er.message = "Electrum-style word list failed verification";
+		return false;
+	}
+
+	namespace po = boost::program_options;
+	po::variables_map vm2;
+
+	{
+		po::options_description desc("dummy");
+		const command_line::arg_descriptor<std::string, true> arg_password = {"password", "password"};
+		const char *argv[4];
+		int argc = 3;
+		argv[0] = "wallet-rpc";
+		argv[1] = "--password";
+		argv[2] = req.password.c_str();
+		argv[3] = NULL;
+		vm2 = *m_vm;
+		command_line::add_arg(desc, arg_password);
+		po::store(po::parse_command_line(argc, argv, desc), vm2);
+	}
+
+	std::unique_ptr<tools::wallet2> wallet = tools::wallet2::make_new(vm2, nullptr).first;
+
+	wallet->set_subaddress_lookahead(0, 0);
+	wallet->set_seed_language(language);
+	wallet->set_refresh_from_block_height(req.refresh_start_height);
+
+	try
+	{
+		if(decode_14)
+			wallet->generate_new(wallet_file, req.password, &seed_14, seed_extra, false);
+		else
+			wallet->generate_legacy(wallet_file, req.password, seed_25, false);
+
+		if(!wallet)
```

Comment:
I do not understand this check: wallet is a unique pointer and use many time with `->` before. If wallet is a `nullptr` tha nwe did many invalid memory accesses before. Could it be that this  check must be in line `2496`

Reply:
I think you are correct and I should move it to `2496`